### PR TITLE
Moving BigTableWriteSchemaTransformProvider to the beam repo

### PR DESCRIPTION
**Pull request by Dippatel98**
Moving BigTable Write Schema Transform to the Beam Repo (DataflowTemplates PR: [#821](https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/821))

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Dippatel98** [comment]
r: @manavgarg, @johnjcasey , @ahmedabu98

**Comment by Dippatel98** [comment]
Run Java_GCP_IO_Direct PreCommit

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BeamSchemaToBytesTransformers.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.primitives.Bytes;
+import org.checkerframework.checker.initialization.qual.UnknownInitialization;
+import org.joda.time.format.ISODateTimeFormat;
+
+@SuppressWarnings("nullness")
```
Comment (johnjcasey) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BeamSchemaToBytesTransformers.java:32]: We should not include new code in the beam repo with this suppression

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigTableIOWriteSchemaBasedTransform.java [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+      if (!(columnFamilies.containsAll(inputColumns)
+          && columnFamilies.size() == inputSchema.getFields().size())) {
+        throw new IllegalArgumentException(
+            String.format(
+                "Unable to match input schema with the columns of the destination "
+                    + "table in Bigtable. Fields missing in BigTable: %s.",
+                inputColumns.removeAll(columnFamilies)));
+      }
+    } catch (IOException e) {
+      // TODO(pabloem): HANDLE THIS POSSIBILITY
```
Comment (johnjcasey) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigTableIOWriteSchemaBasedTransform.java:161]: at the very minimum, we need to rethrow here

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigTableIOWriteSchemaBasedTransform.java [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+                    }))
+            .setCoder(KvCoder.of(ByteArrayCoder.of(), SchemaCoder.of(recordBytesSchema)));
+
+    // STEP 3: Convert KV<bytes, Row> into KV<ByteString, List<SetCell<...>>>
+    PCollection<KV<ByteString, Iterable<Mutation>>> bigtableMutations =
+        byteEncodedKeyedRows.apply(
+            "buildMutations",
+            ParDo.of(
+                new DoFn<KV<byte[], Row>, KV<ByteString, Iterable<Mutation>>>() {
+                  @SuppressWarnings("nullness")
```
Comment (johnjcasey) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigTableIOWriteSchemaBasedTransform.java:279]: we should probably not be suppressing nullness

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BeamSchemaToBytesTransformers.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+          if (row.getDateTime(field.getName()) == null) {
+            return new byte[0];
+          } else {
+            return ISODateTimeFormat.basicDateTime()
+                .print(row.getDateTime(field.getName()))
+                .getBytes(StandardCharsets.UTF_8);
+          }
+        });
+      default:
+        return null;
```
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BeamSchemaToBytesTransformers.java:112]: why not throw the illegal argument exception here? would make BigTableIOWriteSchemaBasedTransform(lines 239-250) more readable

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigTableWriteSchemaTransformConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  @SchemaFieldDescription("The ID of the Bigtable instance to write to.")
+  public abstract String getInstanceId();
+
+  @SchemaFieldDescription("The ID of the Bigtable table to write to.")
+  public abstract String getTableId();
+
+  @SchemaFieldDescription("Columns that make up a row's key.")
+  public abstract List<String> getKeyColumns();
+
+  public abstract @Nullable String getEndpoint();
```
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigTableWriteSchemaTransformConfiguration.java:45]: Missing a `@SchemaFieldDescription` here?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigTableWriteSchemaTransformConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  private static final AutoValueSchema AUTO_VALUE_SCHEMA = new AutoValueSchema();
+  private static final TypeDescriptor<BigTableWriteSchemaTransformConfiguration> TYPE_DESCRIPTOR =
+      TypeDescriptor.of(BigTableWriteSchemaTransformConfiguration.class);
+  private static final SerializableFunction<BigTableWriteSchemaTransformConfiguration, Row>
+      ROW_SERIALIZABLE_FUNCTION = AUTO_VALUE_SCHEMA.toRowFunction(TYPE_DESCRIPTOR);
+
+  /** Transform configuration to a {@link Row}. */
+  public Row toBeamRow() {
+    return ROW_SERIALIZABLE_FUNCTION.apply(this);
+  }
```
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigTableWriteSchemaTransformConfiguration.java:63]: I'm not sure if you need all of this

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigTableWriteSchemaTransformConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    public abstract Builder setInstanceId(String value);
+
+    public abstract Builder setTableId(String value);
+
+    public abstract Builder setKeyColumns(List<String> value);
+
+    public abstract Builder setEndpoint(@Nullable String endpoint);
+
+    public abstract Builder setAppProfileId(@Nullable String appProfile);
```
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigTableWriteSchemaTransformConfiguration.java:77]: It's good practice to set identical value names (ie. projectId, instanceId, etc)

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigTableWriteSchemaTransformProvider.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.beam.sdk.schemas.transforms.SchemaTransformProvider;
+import org.apache.beam.sdk.schemas.transforms.TypedSchemaTransformProvider;
+
+@AutoService(SchemaTransformProvider.class)
+public class BigTableWriteSchemaTransformProvider
+    extends TypedSchemaTransformProvider<BigTableWriteSchemaTransformConfiguration> {
+
+  @Override
+  public String identifier() {
+    return "beam:schematransform:org.apache.beam:bigtable_write:v1";
```
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigTableWriteSchemaTransformProvider.java:33]: There's a BigtableWriteSchemaTransformProvider (for xlang) getting ready to be merged in #27146 with the same URN.

**Review by ahmedabu98 — COMMENTED** [review]
Thanks, left a few comments. 

There's an apparent conflict here with a xlang-oriented BigtableWriteSchemaTransformProvider getting ready for merge in #27146. We should probably resolve into one or make it clear in the naming that they are for different use cases

**Comment by Dippatel98** [comment]
Closing this PR as we decided to go with the BigTableWriteSchemaTransformProvider present in the beam repo (#27146).

**Pull request closed without merging**